### ISSUE TITLE
Review-2018.05.06-杨磊

### ISSUE THREAD

**Issue by DIYer22** [DIYer22/weekly-review#135]
### 一. 写工具库 [Box-X](https://github.com/DIYer22/Box-X)

原本打算花两天完善一下自己的工具库，但越写发觉坑越多，想法也越多，每天都完不成既定目标。导致了 这周没能平衡好时间管理，花了大量的时间精力完善 Box-X，加了不少 feature，增加了 1k 的代码

[![18 05 04 commitsforboxx](https://user-images.githubusercontent.com/10448025/39710302-0626111e-524f-11e8-85f6-5af7ef75485d.png)](https://github.com/DIYer22/Box-X/pulse)

#### 感想：   
 * 软件工程就是花 100% 的时间，完成 90%的工作， 再花 100% 的时间 完成剩下的 90% 的工作。

 * (Python 2/3) * (Python@Shell / IPython@Shell / Qt-IPython) 之间 有很多琐碎的细节不同， 一些很 hack 的 feature 要做到都兼容非常痛苦(比如 涉及到 sys._getframe()、sys.excepthook)
 
 * 预估工作量时，总是过于乐观。完不成又有点不服气， 时间安排就乱了

 * 学习使用[持续集成(Continuous Integration，简称 CI)](http://www.ruanyifeng.com/blog/2017/12/travis_ci_tutorial.html)， CI 是自动化测试与集成代码的技术， 大大减少了代码整合成本。对做项目很有帮助

### 二. 跑 Kaggle 的 [Furniture](https://www.kaggle.com/c/imaterialist-challenge-furniture-2018/data) 

train set 的各类别的训练数据量分布是及不均衡的：   
![18 05 06 train](https://user-images.githubusercontent.com/10448025/39710311-0cd71ae4-524f-11e8-825a-a748a64b0b73.png)   
从 4000 张/类到 300多张/类， 相差了 10 倍 

在 6400 张 validation set 中，则是每个类都挑选了50 张

在 base model 上，预测出来 12,800 test set 的分布如下：   
![18 05 06 test](https://user-images.githubusercontent.com/10448025/39710317-0f2e4bfa-524f-11e8-857d-328b701693ea.png)
预测出来的每一类数目基本在70到120之间， 因此 我估计 test set 每一类都选了100张来测试。

为消除 train 和 test set 的分布不一致的影响，我给 Corss Entropy Loss  加了个权重：   

```Python
weights = 1./counts  # counts是128维的Tensor,第i个数表示第i类的训练集数量
weights = weights/weights.mean()
```
然后 再在最好的 base model 上 Fine-tuning。

结果提交到 Kaggle 上， 点数和 base model 一模一样 (已经确认结果有变化)， 想不太明白改 Loss 为何不 Work。。。

**Comment by ilydouble** [comment]
你做的工作总是超出我的预期，让我很震撼。由于我做云服务，所以我知道现在主流的软件开发已经是DevOps，称为一站式开发，CI应该是其中一部分。如果要我给出建议，我必须花一点时间听你讲一讲xBox然后我上手操作一遍，看你的时间了
第二个工作，你是想针对不均衡的数据通过改loss优化吗，这个权重是写死的吗？还是学习出来的？如果网络足够深，数据量足够大，集成学习对于分类性能的改进就会比较有限。如果设置loss权重，那么你设置的未必好，需要反复测试不同的权重，或者可以利用对抗网络，学习这个loss，当然也可以采用其他的方式尝试合理的设置权重。
最后，不管在任何情况下，多多沟通总是好的，我嗓子说不出话了，有问题微信沟通吧。

**Comment by DIYer22** [comment]
xBox 是微软推出的次世代游戏机，是硬核玩家居家必备.............    
就皮一下😂

Box-X 是我之前的 yllab 发展出来 面向开源的版本， 
用于视觉代码的快速开发和调试   
现在 可以直接通过`pip install boxx`安装了    
兼容 Python2 /3 , Linux / Win (由于我没有mac, OS X 兼容还在开发中)    
我打算 把 Box-X 的文档质量做好一些， 供大家使用    
Box-X 的文档还在紧张制作中， 做好了会第一时间分享给大家   

PS. 老师注意休息，身体健康要紧
